### DISCUSSION TITLE
When does io_uring create tasks or otherwise start iou-wrk threads?

### DISCUSSION THREAD

**Discussion by armanbilge** [axboe/liburing#842] [Q&A]
I am seeking more detailed information about when submitted I/O operations create tasks that are executed on the io_uring worker pool. This [blog post](https://blog.cloudflare.com/missing-manuals-io_uring-worker-pool/) has been a good resource, but it doesn't answer all of my questions.

My motivating "case study":

In https://github.com/armanbilge/fs2-io_uring/issues/56 I created a TCP echo server with `io_uring_prep_recv` and `io_uring_prep_send`. I was surprised to discover that when running it spawns a number of `iou-wrk` threads, which I assumed should not be necessary for non-blocking network I/O.

So in https://github.com/armanbilge/fs2-io_uring/pull/61 I added a `io_uring_prep_poll_add` that then `IOSQE_IO_LINK`s to the `recv`/`send`. After making this change, it appears that my echo server no longer spawns `iou-wrk` threads.

I'd really like to understand why that is and how I can be more conscious in the future about how to submit I/O that does not require the worker pool, if possible. For example, I am wondering if I should make a similar change for `accept`.

Thanks in advance!

**Comment by isilence** [comment]
That shouldn't happen, not often, and it should do fine if you just submit a send/recv without polling first.

What is your kernel version? Is it nonblocking (SOCK_NONBLOCK, MSG_DONTWAIT)? Do you use `MSG_WAITALL`?

**Reply by armanbilge** [reply]
Thanks for the response, that's really helpful information 🤔 so I must be doing something wrong.

Kernel version is 5.15.
```
$ uname -a
Linux armanbilge-fs2iouring-v9qzh8jt6kk 5.15.0-47-generic #51-Ubuntu SMP Thu Aug 11 07:51:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

Currently not using any of `SOCK_NONBLOCK`, `MSG_DONTWAIT`, or `MSG_WAITALL`, is that the issue? I thought using these would possibly result in `EAGAIN` ... been a bit confused how this works, but https://github.com/axboe/liburing/issues/364 seems relevant.

Thanks again, much appreciated!

**Reply by armanbilge** [reply]
Quick follow-up that I tried adding `MSG_DONTWAIT` and this resulted in `EAGAIN`. I'd also tried using `SOCK_NONBLOCK` in https://github.com/armanbilge/fs2-io_uring/pull/57 but that also did not seem to make a difference.

**Reply by isilence** [reply]
Nah, there are not a problem, it'll just make looking up code easier. In general, it should be better without non-blocking flags, it should never block the submitter task / process in either case.

I'll take a look, I might also need a reproducer, it always makes things simpler. Also, would be great if you can run a newer kernel to see whether it's something that has already been fixed.

**Reply by armanbilge** [reply]
> I might also need a reproducer, it always makes things simpler

Yes definitely, I say the same thing to my own users 😜 ok thanks, will check a newer kernel and work on a reproducer.

**Reply by armanbilge** [reply]
Ok, I think this is a reproducer? (excuse my lazy C 😅 ) It's not doing what my original reproducer does (this C version is contrived), but it _does_ make an `iou-wrk` thread appear on kernel 6.2.0.

<details>

```c
#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <stdlib.h>
#include <string.h>
#include <netinet/udp.h>
#include <arpa/inet.h>
#include <netdb.h>

#include <liburing.h>

int main(int argc, char *argv[])
{
  struct io_uring ring;
  io_uring_queue_init(8, &ring, 0);
  int fd = socket(AF_INET, SOCK_STREAM, 0);
  struct addrinfo * ai;
  getaddrinfo("httpbin.org", "80", NULL, &ai);
  connect(fd, ai->ai_addr, ai->ai_addrlen);
  char* req = "GET /get HTTP/1.1\nHost: httpbin.org\n\n";
  
  int i;
  for (i = 0; i < 1024 * 1024; ++i) {
    struct io_uring_sqe *sqe = io_uring_get_sqe(&ring);
    io_uring_prep_send(sqe, fd, req, sizeof(req) - 1, MSG_NOSIGNAL);
    io_uring_submit(&ring);
  }
  sleep(600);
}
```

</details>

**Reply by isilence** [reply]
Yeah, it should be able to easily trigger io-wq. What happens is that after you fill up the socket's tx queues, io_uring will queue the rest requests for polling. Then, when the socket completes some pending IO, it'll wake up all the pending requests and they will rush racing with each other trying to send data. A small part will succeed,  but the rest will get queued for io-wq. In 6.4 it's a bit different, but for 1M ops I guess you can trigger it as well.

**Reply by isilence** [reply]
The problem is though that it doesn't make sense to issue multiple sends to a single TCP socket, the ordering b/w requests will be undefined. Does your original test spam sockets as well?

**Reply by armanbilge** [reply]
> A small part will succeed, but the rest will get queued for io-wq

Hmm, I see, although I don't understand why they don't arm for polling.

> The problem is though that it doesn't make sense to issue multiple sends to a single TCP socket, the ordering b/w requests will be undefined. Does your original test spam sockets as well?

Yeah, it was contrived :) no, my original test does not do that. It submits only one read and one write per socket at a time and waits for the completion. I was having a harder time making a C minimizer for that, but I can try again.

**Reply by isilence** [reply]
> > A small part will succeed, but the rest will get queued for io-wq
> 
> Hmm, I see, although I don't understand why they don't arm for polling.

They do but starting from 6.3 and only a limited number of times (i.e. 128).  In any case perf for that kind of pattern will be awful in any case.

> > The problem is though that it doesn't make sense to issue multiple sends to a single TCP socket, the ordering b/w requests will be undefined. Does your original test spam sockets as well?
> 
> Yeah, it was contrived :) no, my original test does not do that. It submits only one read and one write per socket at a time and waits for the completion. I was having a harder time making a C minimizer for that, but I can try again.

Perfect, thanks

**Reply by armanbilge** [reply]
Ok, here's a better reproducer. It submits one send at a time and waits for the completion before sending the next.

```
Linux armanbilge-sandbox-6mp9censl6d 6.1.27-060127-generic #202305051509 SMP PREEMPT_DYNAMIC Fri May  5 15:48:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ ps -T -p 4693
    PID    SPID TTY          TIME CMD
   4693    4693 pts/3    00:00:35 a.out
   4693    4694 pts/3    00:00:00 iou-wrk-2841766
```

<details>

```c
#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <stdlib.h>
#include <string.h>
#include <netinet/udp.h>
#include <arpa/inet.h>
#include <netdb.h>

#include <liburing.h>

int main(int argc, char *argv[])
{
  struct io_uring ring;
  io_uring_queue_init(8, &ring, 0);
  int fd = socket(AF_INET, SOCK_STREAM, 0);
  struct addrinfo * ai;
  getaddrinfo("httpbin.org", "80", NULL, &ai);
  connect(fd, ai->ai_addr, ai->ai_addrlen);
  char* req = "GET /get HTTP/1.1\nHost: httpbin.org\n\n";
  
  while (true) {
    struct io_uring_sqe *sqe = io_uring_get_sqe(&ring);
    struct io_uring_cqe *cqe;
    io_uring_prep_send(sqe, fd, req, sizeof(req) - 1, MSG_NOSIGNAL);
    
    int ret = io_uring_submit(&ring);
    if (ret != 1) {
      perror("io_uring_submit");
      return 1;
    }
    
    ret = io_uring_wait_cqe(&ring, &cqe);
    if (ret < 0) {
      perror("io_uring_wait_cqe");
      return 1;
    }
    
    io_uring_cqe_seen(&ring, cqe);
  }
}
```

</details>

**Reply by armanbilge** [reply]
Here's another version, which doesn't create an `iou-wrk` thread on 6.1, but it does on 5.15.

<details>

```c
#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <stdlib.h>
#include <string.h>
#include <netinet/udp.h>
#include <arpa/inet.h>
#include <netdb.h>

#include <liburing.h>

int main(int argc, char *argv[])
{
  struct io_uring ring;
  io_uring_queue_init(8, &ring, 0);
  int fd = socket(AF_INET, SOCK_STREAM, 0);
  struct addrinfo * ai;
  getaddrinfo("httpbin.org", "80", NULL, &ai);
  connect(fd, ai->ai_addr, ai->ai_addrlen);
  char* req = "GET /get HTTP/1.1\nHost: httpbin.org\n\n";
  
  while (true) {
    struct io_uring_sqe *sqe = io_uring_get_sqe(&ring);
    struct io_uring_cqe *cqe;
    io_uring_prep_send(sqe, fd, req, strlen(req), MSG_NOSIGNAL);
    
    int ret = io_uring_submit(&ring);
    if (ret != 1) {
      perror("io_uring_submit");
      return 1;
    }
    
    ret = io_uring_wait_cqe(&ring, &cqe);
    if (ret < 0) {
      perror("io_uring_wait_cqe");
      return 1;
    }
    
    io_uring_cqe_seen(&ring, cqe);
  }
}
```

</details>

**Reply by isilence** [reply]
I grabbed a fresh stable 5.15 but couldn't reproduce, not sure what's the difference.
Where did you get the kernel from? Did you compile it yourself or was it provided by a Linux distribution?

**Reply by armanbilge** [reply]
Huh, thanks for trying. It's a 5.15.0-71-generic kernel that I got with an Ubuntu 22.04 LTS server. I just checked again now and I can still reproduce. It took maybe 30 seconds before the `iou-wrk` thread appeared.

**Reply by isilence** [reply]
I was running it for an hour, nothing, but it was qemu. I asked about the kernel because distributions apply random kernel patches, so it might be different.

**Reply by axboe** [reply]
@armanbilge I tried this with 5.15-stable (5.15.117) and with 5.15.0-75-generic and can't seem to reproduce it either. Assuming you updated to -75 as well, does it happen for you there?

**Comment by axboe** [comment]
What kernel is being used?

**Reply by isilence** [reply]
See above, 5.15 and @armanbilge tried out 6.2, which should be before retries and `APOLL_MAX_RETRY` came in.
